client/tinychain/ml/linalg.py

```diff
@@ -117,6 +117,99 @@ def qr_step(cxt, Q: Tensor, R: Tensor, k: UInt) -> Map:
     return Tensor(QR['Q'])[:cxt.n].transpose(), Tensor(QR['R'])[:cxt.n]
 
 
+class PLUFactorization(Map):
+    """
+    PLU factorization of a given `[N, N]` matrix.
+    """
+    
+    @property
+    def p(self) -> Tensor:
+        """
+        Permutation matrix as an `[N, N]` `Tensor`.
+        """
+        return Tensor(self['p'])
+
+    @property
+    def l(self) -> Tensor:
+        """
+        Lower-triangular matrix as an `[N, N]` `Tensor`.
+        """
+        return Tensor(self['l'])
+
+    @property
+    def u(self) -> Tensor:
+        """
+        Upper-triangular matrix as an `[N, N]` `Tensor`.
+        """
+        return Tensor(self['u'])
+
+@post_op
+def plu(x: Tensor) -> PLUFactorization:
+    """Compute the PLU factorization of the given `matrix`.
+    
+    Args:
+        `x`: a matrix with shape `[N, N]`
+
+    Returns:
+        A `[p, l, u]` list of `Tensor`s where
+        `p`: a permutation matrix,
+        `l`: lower triangular with unit diagonal elements,
+        `u`: upper triangular.
+    """
+
+    def permute_rows(x: Tensor, p: Tensor, start_from: UInt) -> Map:
+
+        @post_op
+        def step(p: Tensor, x: Tensor, k: UInt) -> Map:
+            p_k, p_kp1 = p[k].copy(), p[k + 1].copy()
+            x_k, x_kp1 = x[k].copy(), x[k + 1].copy()
+
+            return Map(After(
+                [
+                    p[k].write(p_kp1),
+                    p[k + 1].write(p_k),
+                    x[k].write(x_kp1),
+                    x[k + 1].write(x_k),
+                ],
+                Map(p=p, x=x, k=k + 1)
+            ))
+
+        @post_op
+        def cond(x: Tensor, k: UInt):
+            return (k < UInt(Tuple(x.shape)[0]) - 1).logical_and(x[k, k].abs() < 1e-3)
+
+        return Map(While(cond, step, Map(
+            p=p.copy(),
+            x=x.copy(),
+            k=start_from,
+        )))
+
+    @post_op
+    def step(p: Tensor, l: Tensor, u: Tensor, i: UInt) -> Map:
+        pu = permute_rows(p=p, x=u, start_from=i)
+        u = Tensor(pu['x'])
+        p = Tensor(pu['p'])
+        factor = Tensor(u[i + 1:, i] / u[i, i])
+        print(factor.expand_dims(axis=0).transpose())
```

Comment:
Remove this `print` statement--it's fine to use `print` for debugging but you should use Python's `logging` module if you want any compile-time logs to be accessible to the user